iron-focus-wrap-behavior.html

```diff
@@ -0,0 +1,236 @@
+
+<script>
+(function() {
+'use strict';
+
+/**
+`Polymer.IronFocusWrapBehavior` can wrap (and trap) the focus within an element
+as soon as it gets focused, by listening at `TAB` keyboard events.
+
+When `TAB` is pressed and the `lastFocusable` is already focused, `IronFocusWrapBehavior`
+will move the focus to `firstFocusable`.
+Likewise, when `Shift+TAB` is pressed and
+the `firstFocusable` is already focused, `IronFocusWrapBehavior` will move
+the focus to `lastFocusable`.
+
+    <button id="first">first</button>
+    <content></content>
+    <button id="last">last</button>
+
+    ...
+
+    ready: function() {
+      this.wrapFocus = true;
+      this.firstFocusable = this.$.first;
+      this.lastFocusable = this.$.last;
+    }
+
+In addition, `Polymer.IronFocusWrapBehavior` can be used to save and restore the
+focus of an active element.
+
+    // Before moving the focus, store the deep active element.
+    this.saveDeepActiveElement();
+    this.focus();
+
+    // later on...
+    this.blur();
+    this.restoreDeepActiveElement();
+
+@demo demo/index.html
+@polymerBehavior Polymer.IronFocusWrapBehavior
+*/
+
+  Polymer.IronFocusWrapBehaviorImpl = {
+
+    properties: {
+      /**
+       * The first focusable element. It should be a child of the element or the
+       * element itself. Defaults to `this`.
+       * @type {Element}
+       */
+      firstFocusable: {
+        type: Element,
+        value: function() {
+          return this;
+        }
+      },
+      /**
+       * The last focusable element. It should be a child of the element or the
+       * element itself. Defaults to `this`.
+       * @type {Element}
+       */
+      lastFocusable: {
+        type: Element,
+        value: function() {
+          return this;
+        }
+      },
+      /**
+       * If true, it will wrap the focus on `TAB` or `Shift+TAB` when the focus
+       * is already in the element.
+       */
+      wrapFocus: Boolean
+    },
+
+    ready: function() {
+      this.__deepActiveElement = null;
+      this.addOwnKeyBinding('tab', '__onTab');
+    },
+
+    /**
+     * Updates the `storedDeepActiveElement` to the currently focused element.
+     */
+    saveDeepActiveElement: function() {
+      this.__deepActiveElement = this.__getDeepActiveElement();
+    },
+
+    /**
+     * Restores the focus to the `storedDeepActiveElement`.
+     */
+    restoreDeepActiveElement: function() {
+      if (this.__deepActiveElement) {
+        this.__deepActiveElement.focus();
+      }
+    },
+
+    /**
+     * Searches and sets the `firstFocusable` and `lastFocusable`. This is very
+     * expensive and you're encouraged to set first and last focusable yourself
+     * if you know the content.
+     * @param {boolean=} searchInShadowDOM If true, the search is performed in
+     * the shadow dom.
+     * @protected
+     */
+    _searchAndSetFocusables: function(searchInShadowDOM) {
```

Comment:
wdyt about having a `root` instead of `searchInShadowDOM`?  so If I'd optimize the querySelectorAll to the subtree that has the focusable nodes.  I could also do this:

```js
// light DOM
_searchAndSetFocusables(Polymer.dom(this))  
// Shadow DOM
_searchAndSetFocusables(Polymer.dom(this.root))  
```